exhibitions/app/controllers.js

```diff
@@ -0,0 +1,116 @@
+
+// @flow
+import {List} from 'immutable';
+import {model} from 'common';
+import {filters} from 'common';
+import {services} from 'common';
+const {createPageConfig} = model;
+const {getBreakpoint} = filters;
+const {
+  getPrismicApi,
+  parseBody,
+  asText,
+  asHtml,
+  parsePicture,
+  parseImagePromo,
+  parsePromoListItem
+} = services;
+
+async function getTypeById(req: ?Request, types: Array<DocumentType>, id: string, qOpts: Object<any>) {
+  const prismic = await getPrismicApi(req);
+  const doc = await prismic.getByID(id, qOpts);
+  return doc && types.indexOf(doc.type) !== -1 ? doc : null;
+}
+
+export async function renderExhibition(ctx, next) {
+  const id = `${ctx.params.id}`;
+  const isPreview = Boolean(ctx.params.preview);
+  const exhibitionContent = await getExhibition(id, isPreview ? ctx.request : null);
+  const format = ctx.request.query.format;
+  const path = ctx.request.url;
+  const tags = [{
+    text: 'Exhibitions',
+    url: '/whats-on/exhibitions/all-exhibitions'
+  }];
+
+  if (exhibitionContent) {
+    if (format === 'json') {
+      ctx.body = exhibitionContent;
+    } else {
+      ctx.render('exhibition', {
+        pageConfig: createPageConfig({
+          path: path,
+          title: exhibitionContent.exhibition.title,
+          inSection: 'whatson',
+          category: 'publicprograms',
+          contentType: 'exhibitions',
+          canonicalUri: `${ctx.globals.rootDomain}/exhibitions/${exhibitionContent.exhibition.id}`
+        }),
+        exhibitionContent: exhibitionContent,
+        isPreview: isPreview,
+        tags
+      });
+    }
+  }
+
+  return next();
+}
+
+export async function getExhibition(id: string, previewReq: ?Request): Promise<?ExhibitionContent> {
+  const exhibition = await getTypeById(previewReq, ['exhibitions'], id, {});
+
+  if (!exhibition) { return null; }
+
+  const ex = parseExhibitionsDoc(exhibition);
+
+  const galleryLevel = exhibition.data.galleryLevel;
+  const promoList = exhibition.data.promoList;
+  const relatedArticles = promoList.filter(x => x.type === 'article').map(parsePromoListItem);
+  const relatedEvents = promoList.filter(x => x.type === 'event').map(parsePromoListItem);
+  const relatedBooks = promoList.filter(x => x.type === 'book').map(parsePromoListItem);
+  const relatedGalleries = promoList.filter(x => x.type === 'gallery').map(parsePromoListItem);
```

Comment:
you're looping through the entire promoList 4 times here, perhaps you could do 

```js
const relatedArticles = [];
const relatedEvents = [];
// ...

for (const item of promoList) {
  switch (x.type) {
    case 'article':
      relatedArticles.push(parsePromoListItem(x));
      break;
// ...
```

(less functional but hey, perf!)

Reply:
I think readability would win here - we'd be shaving off negligible amounts in comparison to where we might win, and maybe having something like Prepack would be better instead of writing compiled code.

https://prepack.io/

Please keep pointing these out though!

Reply:
yeah Prepack is very exciting! and I agree about readability + because this is serverside we don't need to worry about perf that much

Reply:
> because this is serverside we don't need to worry about perf that much

More you have to worry about a different type of perf, although most of everything runs on V8 now, so it's much and a muchness.